haystack/pipelines/base.py

```diff
@@ -49,6 +49,11 @@
 from haystack.telemetry import send_event
 from haystack.utils.experiment_tracking import MLflowTrackingHead, Tracker as tracker
 
+try:
+    import ray
+except:
+    ray = None  # type: ignore
+
```

Comment:
Just a suggestion as I don't like the ray dependency and the class name string comparisons too much: We could instead introduce two methods that would be overridden by `RayPipeline` (e.g. `_run_node` and `_get_run_node_signature`)

Reply:
It could be done, but I am not sure that just to avoid the ray dependency and the class name string comparison worth the added code complexity of two more methods. Remember, this is primarily a bug-fix to fix that nasty issue where the `RayPipeline` can only be used for the simplest of pipelines. While fixing that @ZanSara wanted this part of the code to be refactored, but I was told that the whole `RayPipeline` will be refactored in the future.
I will leave it up to @ZanSara to decide whether this is needed or not.

Reply:
In a normal situation I would totally agree with @tstadel: having a Ray import in the base Pipeline code is not nice and I would normally ask exactly for an approach like the one you suggested.

However, there are other factors to consider:
- `RayPipeline` is not a widely used pipeline, so this will impact very few users
- The Ray dependency is optional, so there should be no impact on minimal installations
- Both Pipelines are on the line for a complete refactoring soon. The patch, for as hacky as it is, is straightforward to understand, and we have no other pipeline types, so we don't need generalizations.

If the refactoring was not planned yet, I would totally agree this is too hacky, but the refactoring will come soon so I'd rather spend minimum amount of effort into code that is due to be rewritten anyway..

This said, I don't want to push for it if you @tstadel are not comfortable with the outcome. I will also ping @massi to see how strong his opinion is on this specific case and, if he calls for a more tidy solution, we should go for it.

Reply:
If it's only temporary that's ok, of course. Although having a dependency to subclasses is generally never a good idea. I wouldn't expect to find `RayPipeline`-specific behavior in its subclass. Also when it comes to complexity, two well-named small methods that cleanly separate responsibilities can be easier to understand than the two if-blocks: We at least have a hint by the method name about what they are doing and why the distinction was needed without having to fully understand the surrounding code. If the name of these two methods were kind of similar makes we would also better understand that the two parts in code are closely related.
I agree, let's hear @ZanSara's opinion on that.

Reply:
I am happy to go either way, you guys let me know once decided.

Reply:
All right, to make everybody comfortable with this PR I think I will just go ahead with the changes @tstadel recommended.
Update to this PR will be arriving momentarily.

Reply:
It has been refactored as @tstadel  requested. Resolving this conversation.